Modify source time tuple so that it stays within bounds before sampling from distribution

pseeth opened the pull request:
Previously, if the sampled source_time + event_duration exceeded source_duration, the new source_time would simply be source_duration - event_duration (e.g. the last event_duration seconds of the audio file). While this is somewhat okay and unnoticeable for shorter audio files, it works poorly for longer audio files (e.g. music  stems).

Additionally if a user specifies a distribution like uniform, normal, or trunc_norm, but sets the max or the mean such that it exceeds the duration of the file, the returned sampled source_time becomes a const (of source_duration - event_time). This change makes it so that they are still sampling from the requested distribution but now within safe bounds.

The bulk of this change is in a new function called `_modify_source_time` in `core.py`. 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/justinsalamon/scaper/53)
<!-- Reviewable:end -->

pseeth commented:
source_time tuples should also be modified if the minimum is out of bounds as well. For lack of a better  idea, if the minimum + event_duration > source_duration, then the min is set to 0. So if both the minimum and maximum are out of bounds, then the source_time is sampled between 0 and source_duration - event_duration.

Finally for the normal and truncnorm distributions, the result can still be out of bounds if you set mean to source_duration - event_duration. In these cases, the code at the moment just falls back to the old method of simply setting source_time = source_duration - event_duration.

justinsalamon commented:
I'm more or less following, but it would be good to get some clarifications:

Right now, if you ask for an `event_duration` such that that `source_time` + `event_duration` > `source_duration`, `source_time` is set to `source_duration` - `event_duration`. The logic behind this is as follows:
1. `event_duration` always takes top priority. I.e. we always try to accommodate the user specified event duration, even at the cost of modifying `source_time`.
2. If `source_time` is too large to satisfy `event_duration`, we move it backwards in time under the assumption that we want to keep the modified `source_time` as close as possible to the specified `source_time`.
3. If `source_duration` - `event_duration` is negative, it means there is no way to satisfy `event_duration`. Only in this case do we set `source_time` to `0` and `event_duration` to `source_duration`.

If I understanding correctly, the problem you are encountering is that under the current implementation, you often end up with the same `source_time` = `source_duration` - `event_duration` which is undesired.

As such, what you are proposing is to first calculate the valid range for `source_time` such that `event_time` is respected, and then sample from this valid range. That is, first compute the valid range and then sample, instead of first sampling and then checking if the result is valid.

Is this correct?

justinsalamon commented on scaper/core.py:
```diff
@@ -358,6 +359,72 @@ def _validate_distribution(dist_tuple):
+    elif source_time[0] == 'choose':
+        for i, t in enumerate(source_time[1]):
+            if t + event_duration > source_duration:
+                source_time[1][i] = max(0, source_duration - event_duration)
```
Consider the following case:
* `source_duration = 5`
* `event_duration = 3`
* `source_time[1] = [0, 1, 2, 3, 4, 5]`

This current code will update the array to `[0, 1, 2, 2, 2, 2]`, meaning it's much more likely for the user to get a `source_time` of 2 as opposed to `0` or `1`. Presumably the user wants a value chosen at random from all possible values, i.e. `[0, 1, 2]`, with equal likelihood. 

To address this, should we not remove all duplicates from the modified `source_time[1]` array?

pseeth replied:
That's a good idea, we could just make it `list(set(source_time[1]))`?

justinsalamon replied:
yup

justinsalamon replied:
@pseeth looks like we're still missing the conversion to `list(set(source_time[1]))` ?

pseeth replied:
Done!

justinsalamon replied:
Eek, I believe that's not the right place for this. The current location will result in the loop iterating over different list indices compared to those of source_time[1] which I think leads to erroneous behavior.

I meant we should add  `source_time[1] = list(set(source_time[1]))` *after* the loop which updates each value in the list, i.e. right after line 398 and outside the scope of the for loop (but inside the scope of the elif.

justinsalamon commented:
@pseeth I had a quick look at your changes and I think they're fine, but it seems to me like there's a bigger discussion to be had first about the behavior on scaper when faced with a combination of `event_duration`, `source_time` and `source_duration` that can't be satisfied.

Right now the behavior is:
1. Prioritize `event_duration`, at the cost of `source_time`

But, other options are possible, such as:
2. Prioritize `source_time`, at the cost of `event_duration`.

My inclination is to stick with (1), but I wanted to run it by you first.

A related issue is how we prioritize `event_time` versus `event_duration`. Right now `event_duration` is prioritized such that if `event_time` + `event_duration` > `soundscape_duration`, `event_time` get modified to `soundscape_duration - event_duration`.  But I wonder if that's the right play here? Perhaps it would make more sense to prioritize `event_time` over `event_duration`? This would mean the `event_time` would be chosen sampled from the specified distribution, no issues, and then if it ends up going over the soundscape duration it would just be trimmed. Thoughts?

Would like to get your take on these 2 issues: 
1. `event_duration` vs `source_time`
2. `event_duration` vs `soundscape_duration`.

My inclination with (1) is to keep it as is right now, i.e. prioritize `event_duration`. I'm less clear about (2).

pseeth commented:
Hey @justinsalamon, yeah your initial comment  was correct and that's how I've implemented it. The reasoning is that I don't know (or particularly care) the lengths of all of the audio files in my dataset. The more important thing (at least for my purposes) is that what is generated by Scaper is sufficiently random while still sampling from all the possible data. I agree that `event_duration` should be prioritized over `source_time`. I think prioritizing event_time over event_duration could result in clipped or cut-off audio towards the end of soundscapes.

pseeth commented:
Just wanted to elaborate slightly.

Say a user requests ('uniform', 0, 300) for `source_time`. The source audio files are all of varying length, some are around 10 seconds and some are around 300 seconds and some exceed 300 seconds. Say the requested event duration is 1 second. Here's what currently happens:

1. 10 second source file. The source time is sampled from (0, 300) uniformly, but the probability of getting 10 - 1 = 9 for the start time outweighs any other start time as when sampling from (0,300), getting a number in (10, 300) results in that happening. I think it's reasonable to assume that the user is valuing the _distribution_ over source_time (e.g. they want to draw uniformly from each audio file).
2. 300 second source file, the current behavior is unchanged.
3. > 300 second source file, the current behavior is unchanged.

If a user knows the duration of all of their source files, they can fix it on their end when making an event.

justinsalamon commented:
> I think prioritizing event_time over event_duration could result in clipped or cut-off audio towards the end of soundscapes.

It definitely would. 

Basically there's no way to satisfy both the desired distribution of start times and the desired distribution of event durations because the end of the soundscape forces you to prioritize one over the other: either respect event durations (as it is now) at the expense of the start time distribution, or prioritize the distribution of start times at the cost of even durations, as some events will get clipped if they're toward the end of soundscape. 

If you think it makes sense to keep the current priorities (event duration > event time) I'm fine with that.

pseeth commented on scaper/core.py:
```diff
@@ -358,6 +359,72 @@ def _validate_distribution(dist_tuple):
+        if source_time[2] + event_duration > source_duration:
+            source_time[2] = max(0, source_duration - event_duration)
+        if source_time[1] + event_duration > source_duration:
+            source_time[1] = 0
```
Sure, this makes sense. I think the reason I changed it is because (uniform, 5, 5) throws an error as the high has to be strictly greater than the low in the call to the uniform random generator. So I just said whatever and set it to 0. But I could set it to like (uniform, 5 - eps, 5) or something?

justinsalamon replied:
That feels a little hacky. A better option would be to add a check to see if the lower and upper bounds are identical and if so just return that value, otherwise sample from the distribution.

pseeth commented on scaper/core.py:
```diff
@@ -358,6 +359,72 @@ def _validate_distribution(dist_tuple):
                 'number that is equal to or greater than trunc_min.')
 
 
+def _modify_source_time(source_time, source_duration, event_duration):
```
How about `_ensure_valid_source_time_tuple`?

justinsalamon replied:
not keen on including the term `valid` because scaper already has a notion of tuple validity (a tuple that satisfies the expected syntax), which is unrelated to whether the tuple can be satisfied given the specific constraints of a source file (I purposely avoided `valid` in my suggestions for this reason).

How about `_ensure_satisfiable_source_time_tuple`? It's a mouthful but it's clear.

pseeth replied:
Yeah that's fine and makes sense. I'll edit!

pseeth commented:
Made some edits if you can review @justinsalamon. I added a test case for `_ensure_valid_source_time_tuple_bounds` (a name I just picked, easy to change if you have objections!). I used your review to make the test case, hopefully it's doing what we want now!

justinsalamon commented on scaper/core.py:
```diff
@@ -358,6 +359,78 @@ def _validate_distribution(dist_tuple):
+        if source_time[1] + event_duration > source_duration:
+            source_time[1] = max(0, source_duration - event_duration - eps)
+        if source_time[2] + event_duration > source_duration:
+            source_time[2] = max(eps, source_duration - event_duration)
```
As noted above, I don't think we should use eps

justinsalamon commented on scaper/core.py:
```diff
@@ -358,6 +359,78 @@ def _validate_distribution(dist_tuple):
+        if source_time[1] + event_duration > source_duration:
+            source_time[1] = max(0, source_duration - event_duration)
+        if source_time[3] + event_duration > source_duration:
+            source_time[3] = max(0, source_duration - event_duration - eps)
```
No eps please

justinsalamon commented on scaper/core.py:
```diff
@@ -358,6 +359,78 @@ def _validate_distribution(dist_tuple):
+        if source_time[3] + event_duration > source_duration:
+            source_time[3] = max(0, source_duration - event_duration - eps)
+        if source_time[4] + event_duration > source_duration:
+            source_time[4] = max(eps, source_duration - event_duration)
```
No `eps`

justinsalamon commented on scaper/core.py:
```diff
@@ -1211,25 +1284,34 @@ def _instantiate_event(self, event, isbackground=False,
+            source_time = -np.Inf
+            while source_time < 0:
+                source_time = _get_value_from_dist(modified_source_time)
+                if source_time + event_duration > source_duration:
```
Why is this check still necessary? Assuming the `_ensure...` function makes sure the `source_time` tuple is satisfiable, this test is redundant, right? I'm not keen on having `source_time` modified in multiple different places to fix the same issue, can make the program logic hard to keep track of. Unless there's a good reason to keep this test I think we should remove it.

pseeth replied:
The check is still needed because normal distributions can still result in negative times.

justinsalamon replied:
That's caught by `while source_time < 0:`. But I don't see why we need the `if source_time + event_duration > source_duration:` check.

pseeth replied:
Oh, right you are. I didn't read one line down haha.

justinsalamon commented on tests/test_core.py:
```diff
@@ -425,6 +425,44 @@ def __test_bad_tuple_list(tuple_list):
     __test_bad_tuple_list(badargs)
 
 
+def test_ensure_valid_source_time_tuple():
```
Don't forget to update the name of this test function once we settle on a name for the func it's testing :)

pseeth replied:
Will do!

justinsalamon commented:
@pseeth let's be sure to discuss "open questions" (e.g. naming functions) before they're implemented to avoid redundant code review cycles :)

pseeth commented on scaper/core.py:
```diff
@@ -358,6 +359,78 @@ def _validate_distribution(dist_tuple):
+    _validate_distribution(source_time)
+    old_source_time = deepcopy(source_time)
+    source_time = list(source_time)
+    eps = 1e-1 # epsilon to take care of when low == high.
```
How should this be done? Should we change the distribution to be `const`? Or add some logic into the `_sample_uniform`, `_sample_trunc_norm` etc functions?

pseeth replied:
I changed the distribution to `const` in the current implementation. It doesn't break anything and it seemed like the nicest solution so that sampling source_time is always happening in a similar way.

pseeth commented:
Will do. For some reason, can't comment directly but the `list(set(source_times))` fix isn't as straightforward as I thought. Picture that a user wants to actually weight the draw and puts in `[0, 0, 2, 2, 2, 10]` or something, to give a 3/5s chance of getting 2 and a 2/5s chance of getting 0, and then 10 would be out of bounds, say. The set solution would get rid of this and not warn them.

pseeth commented:
Had a bit of time so I just went ahead and addressed the comments. Instead of using `list(set(...))`, I wrote something that instead prevents duplicate times being added for out of bounds items in choose. Also renamed the function as requested.

pseeth commented on scaper/core.py:
```diff
@@ -358,6 +359,78 @@ def _validate_distribution(dist_tuple):
                 'number that is equal to or greater than trunc_min.')
 
 
+def _ensure_valid_source_time_tuple_bounds(source_time, source_duration, event_duration):
```
Yep, changed it to that.

justinsalamon commented:
> Instead of using list(set(...)), I wrote something that instead prevents duplicate times being added for out of bounds items in choose. 

This workaround has an issue: say the user wants to weight the list as you say, and they provide [0, 1, 2, 3, 3, 3] but the value 3 is out of bounds. Then all the 3s would get adjusted but duplicates would be removed, so now the bias towards 3 is gone. Worse still, you end up with mixed semantics where values below the upper boundary can have duplicates and values above it cannot.

I think it might be cleaner to just specify that any duplicates in "choose" will be automatically removed (and we move the duplicate removal code somewhere upstream) such that all values are sampled uniformly. We could then, if there's need for it, introduce a new distribution tuple (e.g. "discrete") where the user can provide a discrete list of values and corresponding weights (e.g. `[("a", 1), ("b", 1), ("c", 10)]`).

justinsalamon commented:
For example we could change https://github.com/justinsalamon/scaper/blob/master/scaper/core.py#L27 from:
`"choose": lambda x: random.choice(x),`
to:
`"choose": lambda x: random.choice(np.unique(x)),`

pseeth commented:
Bump, if you have time, @justinsalamon.

justinsalamon commented on scaper/core.py:
```diff
@@ -1211,25 +1290,34 @@ def _instantiate_event(self, event, isbackground=False,
+                    '{:s} source time tuple went out of bounds '
+                    'source duration ({:.2f}) and event duration ({:.2f}), modified source time '
+                    'tuple to stay within bounds'.format(
+                        label, source_duration, event_duration),
```
Let's just be sure to update this warning to also print out the updated source_duration tuple

pseeth replied:
See my comment on the previous thing about the warning.

justinsalamon commented on scaper/core.py:
```diff
@@ -358,6 +359,84 @@ def _validate_distribution(dist_tuple):
+
+    # If it's a uniform distribution, tuple must be of length 3, We change the 3rd
+    # item to source_duration - event_duration so that we stay in bounds. If the min
+    # out of bounds, we change it to be source_duration - event_duration - eps.
```
I believe there's no use of eps anymore, we can update the comment

pseeth replied:
Done.

justinsalamon commented on tests/test_core.py:
```diff
@@ -425,6 +425,61 @@ def __test_bad_tuple_list(tuple_list):
 
 
+def test_ensure_satisfiable_source_time_tuple():
+    # Documenting the expected behavior of _ensure_valid_source_time_tuples
```
_ensure_satisfiable_source_time_tuple (not valid)

pseeth replied:
Yep! Fixed. Good catch.

justinsalamon commented on tests/test_core.py:
```diff
@@ -425,6 +425,61 @@ def __test_bad_tuple_list(tuple_list):
+    _test_dist = ('truncnorm', 8, 1, 4, 10)
+    _adjusted, warn = scaper.core._ensure_satisfiable_source_time_tuple(
+        _test_dist, source_duration, event_duration)
+    #rtol = 2e-2 to account for eps = 1e-1 in _ensure_valid_source_time_tuple
```
no eps so no rtol, right? remove comment?

pseeth replied:
Fixed.

pseeth commented on scaper/core.py:
```diff
@@ -358,6 +359,84 @@ def _validate_distribution(dist_tuple):
+    
+    source_time = tuple(source_time)
+    # check if the source_time changed from the old_source_time to throw a warning.
+    warn = (source_time != old_source_time)
```
I updated this but had to actually add two warnings - one for if the tuple wasn't valid and the other where if the tuple was valid, it still sampled from outside the bounds (e.g. a normal distribution with the upper bound being the end of the audio file with a std dev of 2 fails half the time).

pseeth replied:
I also added a little note about the "warn" variable here for future people. :)

pseeth commented on scaper/core.py:
```diff
@@ -1307,14 +1309,30 @@ def _instantiate_event(self, event, isbackground=False,
+
         else:
             source_time = 0.0
 
```
I'm open to editing the second warning a bit. Let me know of better text and I'll toss it in there, unless you think this is alright!
